### ISSUE TITLE
Option to play video stream without Piped proxy

### ISSUE THREAD

**Issue by ARBoyGo** [TeamPiped/Piped#2817]
### Describe the feature

In libretube we can turn off Proxy for video stream. 

### Why would this be useful to add?

In current state, It' s useful and we just need to turn on our VPN. 

### Acknowledgements

- [X] I have searched the existing issues and this is **NOT** a duplicate or related to another open issue.
- [X] I have written a short but informative title.
- [X] I filled out all of the requested information in this form.

**Comment by ARBoyGo** [comment]
I was checking more on it. And saw this. https://github.com/TeamPiped/Piped/issues/794#issuecomment-1032487311

**Comment by Bnyro** [comment]
That linked issue is about something different.

**Comment by Bnyro** [comment]
This issue is reasonable in my opinion due to the recent YouTube rate limiting waves.

**Comment by deroverda** [comment]
> > Noob here. Is it YouTubes fault that piped is not working more than it is actually working?
> 
> 
> 
> Assuming that i understand your question the right way, yes, it's Youtubes fault that piped is not working because they are rate limiting very heavily (by piped server IPs). Also, it's Youtubes fault that we even need piped :)
> 
> 

Yes you understood my question. Damn, bloody YouTube. I wonder what will happen now then if they continue with this rate limiting bullocks.

**Comment by chev2** [comment]
I'm assuming such an option would be toggleable per-user and not a global option for the instance owner to change, right?

Also, on top of the benefits mentioned already, this option could additionally help users on instances with heavy traffic load as they would not have to deal with slow video delivery times (I've noticed some instances are pretty slow to buffer videos, likely because of this proxying). Furthermore it could help instance owners offload bandwidth usage, even if by a little bit.

**Comment by Bnyro** [comment]
Yes, the option would be a user preference in the Piped frontend - the proxy will stay enabled by default because probably not all users use a VPN by default. By using Piped you expect to not send any requests directly to YT anymore, and that's how it should stay (if not changed explicitly).
And if added, there'll probably also be a privacy warning when enabling.

**Comment by Bnyro** [comment]
Nevertheless, @FireMasterK has the last word on whether we implement an option to disable the proxy, or not.

Also, videos from YouTube music probably don't work when the proxy is disabled, which should be noted here.

I would work on this if Kavin agrees.

**Comment by FireMasterK** [comment]
> Nevertheless, @FireMasterK has the last word on whether we implement an option to disable the proxy, or not.
> 
> Also, videos from YouTube music probably don't work when the proxy is disabled, which should be noted here.
> 
> I would work on this if Kavin agrees.

I'm already working on this. This cannot be done without an extension, as we have to deal with `CORS`. I'm working on adding support for it in https://github.com/TeamPiped/piped-redirects. Maybe this functionality could also be added to Privacy Redirects too?

**Comment by deroverda** [comment]
So the reason https://piped.kavin.rocks hasn't worked (videos don't play) for many days now is because of YouTube? I just want to know so I know its not something that is wrong on my end. Is this instance also not working for everyone else? Should I switch instance or are none working?

**Comment by Bnyro** [comment]
Yes, see #2809.

**Comment by Qaz-6** [comment]
It is possible to run this feature without a VPN.However, it doesn't open all videos both with and without VPN.🤔

**Comment by felschr** [comment]
> I'm already working on this. This cannot be done without an extension, as we have to deal with `CORS`. I'm working on adding support for it in https://github.com/TeamPiped/piped-redirects. Maybe this functionality could also be added to Privacy Redirects too?

Perhaps Piped could add an option to use the official YouTube embed iframe.
It does provide an API for play, pause, stop, seek, volume, mute, etc., so most of the functionality of piped should:
https://developers.google.com/youtube/iframe_api_reference

Downsides:
- ads can't be removed ([docs](https://support.google.com/youtube/answer/132596?hl=en))
- more Google tracking
- not all of the YT embed UI elements can be hidden

This option should include a warning to explain the risks, but it's probably the only feasible Proxy alternative that doesn't require an extension.

It's still a better compromise for users that would otherwise fall back to using youtube.com instead when the proxy is down.

**Comment by ARBoyGo** [comment]
That's not needed IMO, Cause Invidious has option to turn off proxy.

**Comment by ARBoyGo** [comment]
There is a reason why, I recommend this feature, Because not many people care about privacy they just wanna get rid of Ads. Having proxy makes loading the video slower.

**Comment by felschr** [comment]
Interesting, it does indeed seem to work on Invidious.
I've just tried this on an instance and it connects to:
https://rr5---sn-5hne6nsy.googlevideo.com/videoplayback

**Comment by felschr** [comment]
Here is Invidious' code that builds the URL:
https://github.com/iv-org/invidious/blob/0e4d3d89fc96f808439a72e32a6d783d96f2c429/src/invidious/routes/video_playback.cr#L5-L43

**Comment by Qaz-6** [comment]
> > I'm already working on this. This cannot be done without an extension, as we have to deal with `CORS`. I'm working on adding support for it in https://github.com/TeamPiped/piped-redirects. Maybe this functionality could also be added to Privacy Redirects too?
> 
> Perhaps Piped could add an option to use the official YouTube embed iframe. It does provide an API for play, pause, stop, seek, volume, mute, etc., so most of the functionality of piped should: https://developers.google.com/youtube/iframe_api_reference
> 
> Downsides:
> 
> * ads can't be removed ([docs](https://support.google.com/youtube/answer/132596?hl=en))
> * more Google tracking
> * not all of the YT embed UI elements can be hidden
> 
> This option should include a warning to explain the risks, but it's probably the only feasible Proxy alternative that doesn't require an extension.
> 
> It's still a better compromise for users that would otherwise fall back to using youtube.com instead when the proxy is down.

Ads can't be removed?

**Comment by Qaz-6** [comment]
> > I'm already working on this. This cannot be done without an extension, as we have to deal with `CORS`. I'm working on adding support for it in https://github.com/TeamPiped/piped-redirects. Maybe this functionality could also be added to Privacy Redirects too?
> 
> Perhaps Piped could add an option to use the official YouTube embed iframe. It does provide an API for play, pause, stop, seek, volume, mute, etc., so most of the functionality of piped should: https://developers.google.com/youtube/iframe_api_reference
> 
> Downsides:
> 
> * ads can't be removed ([docs](https://support.google.com/youtube/answer/132596?hl=en))
> * more Google tracking
> * not all of the YT embed UI elements can be hidden
> 
> This option should include a warning to explain the risks, but it's probably the only feasible Proxy alternative that doesn't require an extension.
> 
> It's still a better compromise for users that would otherwise fall back to using youtube.com instead when the proxy is down.

This is not a good idea. It would be great to use YouTube.js 🤔instead.https://github.com/LuanRT/YouTube.js

**Comment by felschr** [comment]
With #3658 this seems more relevant now.

**Comment by NintendoManiac64** [comment]
> With [#3658](https://github.com/TeamPiped/Piped/issues/3658) this seems more relevant now.

Indeed and it's a key reason I did a search to see if anyone already suggested the option.

Normally I prefer invidious for its design that rejects the "javascript all of the things!" asthetic, but invidious previously rejected implementing DeArrow and so I have to bounce between piped and invidious depending on what I'm doing (from what I can tell, DeArrow is not available as a XUL extension for browsers that don't subscribe to the theory of a "Chromified web")
